### PR TITLE
Promote `avoid-recursive-trap` experiment

### PR DESCRIPTION
**Pull request by triarius**
### Description

This has been operating for some time in our dogfood stacks and we have not
noticed any issues. It is time to promote experiment. I have not added any
code to invert the experiment, as I believe it is safe to promote without
that.

### Context

https://3.basecamp.com/3453178/buckets/27608512/messages/6299400220

### Testing
- [x] Tests have run locally (with `go test ./...`). Buildkite employees may check this if the pipeline has run automatically.
- [x] Code is formatted (with `go fmt ./...`)

### CODE REVIEW

File: internal/experiments/experiments.go [hunk 1]
```diff
@@ -30,16 +30,16 @@ const (
...
-	FlockFileLocks    = "flock-file-locks"
-	GitMirrors        = "git-mirrors"
-	InbuiltStatusPage = "inbuilt-status-page"
-	JobAPI            = "job-api"
+	ANSITimestamps     = "ansi-timestamps"
+	FlockFileLocks     = "flock-file-locks"
+	GitMirrors         = "git-mirrors"
+	InbuiltStatusPage  = "inbuilt-status-page"
+	JobAPI             = "job-api"
+	AvoidRecursiveTrap = "avoid-recursive-trap"
```
Comment (DrJosh9000) [inline internal/experiments/experiments.go:42]: why no lexicographic ordering

File: internal/job/executor.go [hunk 2]
```diff
@@ -1032,8 +1007,7 @@ func (e *Executor) defaultCommandPhase(ctx context.Context) error {
 		e.shell.Promptf("%s", cmdToExec)
 	}
 
-	err = e.shell.RunWithoutPrompt(ctx, cmd[0], cmd[1:]...)
-	return err
+	return e.shell.RunWithoutPrompt(ctx, cmd[0], cmd[1:]...)
```
Comment (DrJosh9000) [inline internal/job/executor.go:1010]: This change breaks the `defer` near the start of the function (introduced in 3eaf791):

```go
	var err error
	defer func() { span.FinishWithError(err) }()
```

If we want to do this cleanup, we'd need to make `err` a named return instead.

Comment (triarius) [reply internal/job/executor.go:1010]: I think the same thing will happen here: https://github.com/buildkite/agent/blob/745645930b7d322eb9fc3660a0bb52b9edc945cd/internal/job/executor.go#L683 and possibly several other places yeah?

I think using named return values is an option to fix it: https://goplay.tools/snippet/yHRT8QgN8BG. Then this code can stay as is.

File: internal/experiments/experiments.go [hunk 3]
```diff
@@ -25,42 +25,42 @@ const (
 	// Available experiments
 	AgentAPI                   = "agent-api"
 	DescendingSpawnPrioity     = "descending-spawn-priority"
+	IsolatedPluginCheckout     = "isolated-plugin-checkout"
```
Comment (DrJosh9000) [inline internal/experiments/experiments.go:28]: thank you 🙏

File: internal/job/executor.go (hunk 2 above)
Comment (triarius) [reply internal/job/executor.go]: And I think this is the relevant part of the spec:

> A "return" statement that specifies results sets the result parameters before any deferred functions are executed.

which appears under: https://go.dev/ref/spec#ReturnStmt

**Review by DrJosh9000 — APPROVED** [review]
LGTM 🌿

File: internal/job/executor.go (hunk 2 above)
Comment (DrJosh9000) [reply internal/job/executor.go]: Yep, I think we're on the same page there.

**Comment by torarnv** [comment]
Has the experiment been promoted? I still see the `trap` unless I run with `avoid-recursive-trap` ?

**Comment by torarnv** [comment]
Never mind, I upgraded the agent 3.59.0 -> 3.69.0 and now it's on by default. Sorry :)

**Pull request merged**